### PR TITLE
Deploy Consul, Controller, Kafka, Zookeeper and Invoker in Kubernete

### PR DESCRIPTION
**Pull request by DanLavine**
Deploy Consul and seed the database.
Add all group-vars so that the properties file can be generated.
properly seed Kafka with correct invoker topics
Update deploy script to only deploy couch DB if it does not exist.
Bump deployment time for travis
This PR will check off everything but Nginx from this [issue](https://github.com/openwhisk/openwhisk-devtools/issues/14)

### CODE REVIEW

**Comment by animeshsingh** [comment]
@DanLavine the Travis failing is because of the resource constraint on Travis machine you brought up?

**Comment by DanLavine** [comment]
@animeshsingh No I fixed that issue. The docker-compose ci keeps timing out after 50 min.

**Comment by ddragosd** [comment]
the build is indeed timing out b/c the whisk controller doesn't come up. I'm taking a look at this right now.

**Comment by ddragosd** [comment]
PR #23 opened to fix the issue.

**Comment by DanLavine** [comment]
Should also mention that this PR will also fix the master branch of Kube on Travis. It broke when I updated the image for deploying OpenWhisk on Kube that the master branch uses. This PR bumps the timeout duration which should be long enough for deploying everything.

**Comment by ddragosd** [comment]
thanks for mentioning @DanLavine . Would you then merge with the other branch from #23 in order to have the build pass ? We could end up merging only this PR. WDYT ?

**Comment by animeshsingh** [comment]
@markusthoemmes can we get these in, given @csantanapr is out and these PRs are not in the way of anything core OpenWhisk

**Comment by DanLavine** [comment]
@ddragosd. I would just leave everything split into the 2 PRs. It is kind of odd working in this github/docker space with having to do everything via PRs. All this code is technically in that docker image that is failing CI. It is more of a track of what is currently in the Docker image and how to rebuild it. So with that, I kind of like leaving them as logical separations and it highlights what a few of the pain points about the current process. Leaves it open to try and come up with better approaches for this type of work.

**Comment by ddragosd** [comment]
@DanLavine I've merged #23 in master.

**Comment by DanLavine** [comment]
@ddragosd Thanks. If this branch also gets merged then CI should go green. Also, all of this work is adjacent to docker compose so if something fails after the merge for Kube, I will take a look at it.

**Comment by ddragosd** [comment]
@DanLavine so you're saying it's ok to merge it from your point of view ?

File: kubernetes/ansible-kube/environments/kube/files/invoker.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        hostPath:
+          path: "/var/run/docker.sock"
+      - name: apparmor
+        hostPath:
+          path: "/usr/lib/x86_64-linux-gnu/libapparmor.so.1"
+
+      containers:
+      - name: invoker
+        imagePullPolicy: IfNotPresent
+        image: danlavine/whisk_kube_invoker
```
Comment (ddragosd) [inline kubernetes/ansible-kube/environments/kube/files/invoker.yml]: did you also experiment with https://hub.docker.com/r/openwhisk/invoker/ ?

**Comment by ddragosd** [comment]
Having contributed to the DCOS packages for Mesos deployment I'm a little unsure on the Ansible role for Kube deployments. In Mesos we were able to get the OW packages initialized just fine without Ansible; it's true that we ended up creating a CouchDB docker image with all the initial DBs + Views initialized.

File: kubernetes/ansible-kube/environments/kube/files/invoker.yml (hunk 1 above)
Comment (DanLavine) [inline kubernetes/ansible-kube/environments/kube/files/invoker.yml]: The issue with this is that each invoker needs to know which index they are. In the Kube world, I am doing this by using a couple of things. The Invokers are deployed via [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) which give each instance a static hostname of `invoker-X`. From here I have a custom docker file for the [invoker](https://github.com/DanLavine/openwhisk-devtools/blob/kube/kubernetes/docker/invoker/Dockerfile) which uses this hostname when starting the invoker instance [here](https://github.com/DanLavine/openwhisk-devtools/blob/kube/kubernetes/docker/invoker/invoker_start_script.sh).

Comment (DanLavine) [inline kubernetes/ansible-kube/environments/kube/files/invoker.yml]: This StatefulSet concept is also used to populate the Kafka topics. By using the number of invokers deployed, it generates how may topics should be generated when seeding Kafka.

**Comment by DanLavine** [comment]
@ddragosd This should be fine to merge from my point of view.

File: kubernetes/ansible-kube/environments/kube/files/invoker.yml (hunk 1 above)
Comment (ddragosd) [inline kubernetes/ansible-kube/environments/kube/files/invoker.yml]: Nice. We took a similar approach for DCOS Packages. See https://github.com/openwhisk/openwhisk-devtools/pull/20/files#diff-673d38070852220704426f492fdb101cR95 

Docker command is
```
/invoker/bin/invoker `hostname | tr -dc '0-9'` >> /dev/stderr
```
In this way we can actually use the official `openwhisk/invoker` docker image.

Can this Docker `CMD` option be edited in Kube deployment as well, for those that want to reuse the official image ?

Comment (DanLavine) [inline kubernetes/ansible-kube/environments/kube/files/invoker.yml]: Now that you mention it, I actually might be able to do something similar. I was having some trouble setting the command properly for Kube, but I solved it when setting the command to download all of the images required for the invoker. Let me go back and try to inline everything on this [Kube file](https://github.com/DanLavine/openwhisk-devtools/blob/kube/kubernetes/ansible-kube/environments/kube/files/invoker.yml).

**Comment by DanLavine** [comment]
@ddragosd Updated the Invoker deployment to use the public OpenWhisk image and it passed CI.

**Comment by animeshsingh** [comment]
@ddragosd possible for you to merge this PR?

**Comment by ddragosd** [comment]
@animeshsingh I was waiting to hear back from @DanLavine in regards to the custom `invoker` image to see if there's a chance to reuse the official image instead.  see the comments above.

**Comment by ddragosd** [comment]
@DanLavine 👍 for managing to reuse the official `openwhisk/invoker` image. 

@mrutkows thanks for helping with the merge ! 💯

**Comment by animeshsingh** [comment]
Why the build is still failing? Also @ddragosd  previous question for docker-compose work, can it work across multiple hosts, with distributed components, or its currently targeted for single host deployment?

**Comment by ddragosd** [comment]
> previous question for docker-compose work, can it work across multiple hosts, with distributed components, or its currently targeted for single host deployment?

@animeshsingh this is rather targeted for local deployment and development.

**Pull request merged**